### ISSUE TITLE
Disable macro heuristic for inherited macros

### ISSUE THREAD

**Issue by gkossakowski** [sbt/sbt#1142]
At the moment, any change (even whitespace change) to a file containing a macro will trigger invalidation of files that depend on it. The reason is that we don't have any means to detect change to implementation of a macro that could affects clients that call it so we err on conservative side.

However, the definition of "containing a macro" here means both defines a macro directly in source file or _inherits_ it. Obviously, if we just inherit a macro then the heuristic doesn't need to kick in because we know that macro hasn't been modified.

The way invalidation logic was implemented internally made it difficult to distinguish between defining a macro in source file or inheriting it. The name hashing algorithm is much more amenable to such optimizations so we should consider doing it.

It would help with unexpected recompiles due whitespace change as observed in [scala compiler](https://groups.google.com/forum/#!topic/scala-internals/STmaXmH2RQg/discussion] or in Scala IDE (that inherits from Global).

**Comment by gkossakowski** [comment]
Actually, I realized that the changes to name hashing are not related to this problem. Both old and new algorithms can be improved when it comes to handling of inherited macros. The decision whether given file (API in sbt's terminology) is made here in [HasMacro](https://github.com/sbt/sbt/blob/0.13/compile/api/src/main/scala/xsbt/api/APIUtil.scala#L21) API visitor.

The fix would consist of changing that visitor to not visit inherited members. The rest of the incremental compilation algorithm should work ok with restricted definition of "contains a macro".

When it comes to testing for this issue, I believe that scripted test would be enough. We don't really have unit tests related to `HasMacro` functionality.

**Comment by Duhemm** [comment]
Hi Greg,

Can you confirm that [the test I wrote](https://github.com/Duhemm/sbt/commit/3fcf91709d01ac2b713c2af38a1609e56c88336c) is relevant to this problem ? 

Are you sure that fixing the API visitor should be enough ? I've been able to fix it so that a class that do not explicitly define a macro is not considered as having a macro (just like `Client` in my test), but unfortunately this doesn't seem to be enough. ([Here are the changes](https://github.com/Duhemm/sbt/commit/6625dec6c61173537e1d22d5a5e1c9a93cfd8f29)). 

I will continue investigating this week, but if you have any idea why recompilations are still triggered, please let me know !

(Do you know how I can turn this issue into a pull request ? It would be easier to track changes and discuss it. Should I create a new PR ?)

**Comment by gkossakowski** [comment]
The test looks pretty close but you need a bit more involved structure and changes to other file. Let me demonstrate with an example. Let's say we have the following dependencies

```
Provider <- Client (via inheritance)
Foo -> Client (via composition)
```

The issue described in this ticket is that a whitespace change to _Client_ will cause recompilation of _Foo_ because incremental compiler will pessimistically assume that the macro implementation in Client has changed. That is obviously not possible, because macro implementation is in Provider and we want to detect that.

To sum it up:
- whitespace change in Provider should cause recompilation of both Client and Foo in one iteration (let's test that) 
- whitespace change to Client should not cause recompilation of Foo

Let me know if that clarifies the situation.

**Comment by Duhemm** [comment]
Thank you, this is much clearer ! Is [this test](https://github.com/Duhemm/sbt/commit/7db9979da71f1df3f62fdc7f032938593642a444) aiming at the right issue ?

**Comment by gkossakowski** [comment]
Yes, the test looks good now. It would be great if you could restructure your commits to be in the following order:
- add the test as `pending` (rename `test` file to `pending`) which documents current behavior described in this ticket.
- implement your fix that overrides `visitStructure0` _and_ mark previously added test as passing by renaming `pending` to `test` file

This way we clearly will see what was the previous behavior and how your changes fixes it.
